implemented functions to control stepper motors for linear and rotary units

Ferran-RMS opened the pull request.

Commit:
Added function intended to govern the rotary unit

Inputs: Number of pulses and Direction of rotation (not implemented yet)

Outputs: One GPIO pin

Commit:
Added a variable to take into account the Gear reduction for the pulse count

Commit:
Added a function to select the operation mode

The pins 30 and 31 of the Maple mini will be used to switch in between
modes of operation

Commit:
Code cleanup

Commit:
Added the gear reduction factor and renamed the functions related to the motor

Commit:
Correct variable to display the number of pulses

Commit:
Merged the functions motor_rt and motor_mode (Case 1: clock/dir mode works)

The cases 2,3 and 4 are being debugged

Commit:
Added conditions to introduce the right parameters at each case

Commit:
Cleaned-up the code, as the pins will be used as output anyway

Commit:
Added the variable "start" and setup of the variable

Commit:
Clarified the information for the user about how the parameters must be introduced

Commit:
Added the start/enable signal for the homing mode

Commit:
Changed the types of the loop index and the num_pulses in order to store the required quantity

Commit:
Removed the variable "start" as the modes require it anyway to work

Commit:
Added 200ms of sleep time for the motor controller to be able to read the start signal

Commit:
Updated the pulse calculation

Commit:
Increased the sleep time, in order to allow the motor controller to have enough time to sense the Start signal

Commit:
added function for the mode 3 (similar to the mode 4)

Commit:
Function to control the Linear Unit with an stepper motor

Commit:
debugging, comments

ft commented on firmware/main.c:
```diff
@@ -709,11 +713,16 @@ static void cmd_motor_rt(BaseSequentialStream *chp, int argc, char *argv[]){
 
 
         }
+//	palClearPad(pinPorts[pOpt1].gpio, pinPorts[pOpt1].pin); // Reset pins as its default state is high
```
Personally, I really dislike  commented out code. Why's this commented out? If it's optional, it should be made like that, at compile time or at run time, whichever is preferred.

ft commented on firmware/main.c:
```diff
@@ -834,6 +834,148 @@ static void cmd_motor_mode(BaseSequentialStream *chp, int argc, char *argv[]){
+        bool out1 = false, out2 = false; // Variables to store the state of the outputs
+
+        // Parsing
+        for(i = 0; i < argc; i++) {
```
This looks like it should use getopt(3).

dpiegdon replied:
there is no getopt in chibios, and i'd like to not use newlib. firmware already is rather large (50k bin) also the original codebase from chibios does the same, so this is just style mimicri.

ft commented on firmware/main.c:
```diff
@@ -834,6 +834,148 @@ static void cmd_motor_mode(BaseSequentialStream *chp, int argc, char *argv[]){
+
+        /* Pin definition for the pulse and direction outputs*/
+
+        for(i = 0; i < sizeof(pinPorts)/sizeof(pinPorts[0]); i++) {
```
All these "if-in-loop" sections look very similar, which is usually a sign for missing abstractions. That should make these more readable too.

ft commented on firmware/main.c:
```diff
@@ -615,7 +615,7 @@ static void cmd_motor_rt(BaseSequentialStream *chp, int argc, char *argv[]){
 	palClearPad(pinPorts[start].gpio, pinPorts[start].pin);
 
-	num_pulses = atoi(pOpt) * gear_reduction; // Gets the right number of pulses to be done according to the gear reduction
+	num_pulses = atoi(pOpt) * gear_reduction * 2; // Gets the right number of pulses to be done according to the gear reduction, multiplied by 2
```
There should be some description as to *why* this needs to be times-2. Also, style-nit, when lines get as long as this, the comment should go *before* the code and be made into a proper paragraph-style comment.

ft commented on firmware/main.c:
```diff
@@ -633,6 +633,7 @@ static void cmd_motor_rt(BaseSequentialStream *chp, int argc, char *argv[]){
                         }
 			
 			chprintf(chp, "Pin to use: %d pinPorts size of: %d and number of pulses to produce %d \r\n",pin2use,i,num_pulses);
+			chThdSleepMilliseconds(200); // In order to have enough time to have a pulse for the start/enable signal
```
Again, I'd put the comment in a paragraph-style comment before the code.

ft commented on firmware/main.c:
```diff
@@ -553,17 +553,17 @@ static void cmd_motor_rt(BaseSequentialStream *chp, int argc, char *argv[]){
                         if(++i >= argc) continue;
                         mOpt = argv[i];  // Selects the operation mode
                 }
-		else if(strcmp(argv[i], "-s") == 0) {
```
Again, in this commit, there is a bunch of commented out code. Like I wrote before, if it's obsolete, remove it; if it's optional make it optional.

ft commented on firmware/main.c:
```diff
@@ -555,8 +555,13 @@ static void cmd_motor_rt(BaseSequentialStream *chp, int argc, char *argv[]){
+		|| ((strcmp(mOpt, "2") == 0) && (!pOpt || !dOpt)) 
+		|| ((strcmp(mOpt, "3") == 0) && !sOpt)
+		|| ((strcmp(mOpt, "4") == 0) && !sOpt)
+|| !((strcmp(mOpt, "1") == 0) || (strcmp(mOpt, "2") == 0) || (strcmp(mOpt, "3") == 0) || (strcmp(mOpt, "4") == 0))
```
The structure here looks a little funky. Long line, and closing paren right in front of the goto?

ft commented:
Took a quick look and left a number of comments. I'm not quite familiar with the code base, so lots of my comments are style nits like "Why would you commit commented out code, when you have a revision control system to track changes?". There's also one instance, where I'm pretty sure you'd want to invest a little more time coming up with a proper abstraction in order to remove a sequence of loops, that look very copy-and-pasty.

dpiegdon commented:
as a side note: i created the pull request with ferran to have the option to review the code. this was by no means meant to be commited right away. should have mentioned that.
I have started reviewing and fixing locally now.

Commit:
Code cleanup

Commit:
Pulse calculation update.

The PID calibration of the motor controller was not completely done
(3 Stages), and there was an exact deviation, that is why it was
multiplied by 2.

Commit:
Removal of the function "cmd_motor_mode" as is not longer used

The function was toggling two pins according to the selected case, which at
the moment, this is done in either "motor_rt" or "motor_lu" functions

Commit:
Function cleanup, comment update and removal of unused variables

Commit:
Linear unit function update (C5-E motor controller).

The clock/direction mode is fully operational, including the automatic brake.

Commit:
Code clean-up

Commit:
Corrected the printf of motor_rt function

To show the correct amount of pulses generated given the reduction gear.

Commit:
Linear unit function clean-up

Commit:
Homing function implementation

Commit:
Code clean-up and input parameter protection

Commit:
The mode 2 is only using the direction pin, therefore, they do not need to
be configured unless mode one 1 selected.

If a pin is cleared before having been previously set, the Maple Mini crashes.

Commit:
WIP - Added a function to generate up and down ramps.

In order to smooth the motor movement.

dpiegdon commented:
i finally started reviewing the code.
including my changes, these can be found in branch Ferran-RMS-cleanup.

dpiegdon commented:
Ferran, can you take a look into the mentioned branch, Ferran-RMS-cleanup?
I have a question on cmd_motor_rotary() when using mode 1 or 2:
don't I need to set the pinEnable before, and clear it after executing the pulse train?
this was not done in the older version you wrote, but if I had to guess, I'd say we need this.
I marked the potential location with 'FIXME's.

Ferran-RMS commented:
Hi David !
Is not needed, as in modes 1 and 2, is just waiting for a clk and direction signal in order to move.
The modes 3 and 4 require a enable signal only.

On the SMCi35, the so called "profiles" are just a description of what the motor controller should do on each of them. If a certain movement is programmed, then the motor controller only needs a trigger signal to execute the movement.

Just let me know:)

Cheers!

dpiegdon commented:
ok!
